fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java

```diff
@@ -644,12 +649,22 @@ public LogicalPlan visitCTE(CTERelation node, ExpressionMapping context) {
     @Override
     public LogicalPlan visitSubquery(SubqueryRelation node, ExpressionMapping context) {
         LogicalPlan logicalPlan = transform(node.getQueryStatement().getQueryRelation());
+        OptExpression subQueryOptExpression = logicalPlan.getRoot();
+
         OptExprBuilder builder = new OptExprBuilder(
                 logicalPlan.getRoot().getOp(),
                 logicalPlan.getRootBuilder().getInputs(),
                 new ExpressionMapping(node.getScope(), logicalPlan.getOutputColumn()));
 
         builder = addOrderByLimit(builder, node);
+
+        // store opt expression to ast map if sub-query is project or union
+        OperatorType operatorType = subQueryOptExpression.getOp().getOpType();
+        if (optToAstMap != null && (operatorType == OperatorType.LOGICAL_PROJECT ||
```

Comment:
why we should process project and union?

Reply:
support for subquery and union all's subquery, see `MaterializedViewTextBasedRewriteTest`